File: award.py
class Award(object):
    def __init__(self, name=None, expires_in=None, quality=None):
        self.name = name
        self.expires_in = expires_in
        self.quality = quality

    def update_quality(self, awards):
            for award in awards:
                # Blue Distinction Plus awards should not change
                if award.name != 'Blue Distinction Plus':
                    if award.name == 'Blue First':
                        self.update_bf(award)
                    elif award.name == 'Blue Compare':
                        self.update_bc(award)
                    elif award.name == 'Blue Star':
                        self.update_bs(award)
                    else:
                        self.update_gen(award)
                    award.expires_in -= 1
                    self.fix_overages(award)

    # update Blue Compare award
    def update_bc(self, award):
        if award.expires_in > 0:
            award.quality += 1
        if award.expires_in <= 10:
            award.quality += 1
        if award.expires_in <= 5:
            award.quality += 1
        if award.expires_in <= 0:
            award.quality = 0

    # update Blue First award
    def update_bf(self, award):
        award.quality += 1
        if award.expires_in <= 0:
            award.quality += 1

    # update Blue Star award
    def update_bs(self, award):
        if award.expires_in > 0:
            award.quality -= 2
        else:
            award.quality -= 4

    # update all other awards
    def update_gen(self, award):
        award.quality -= 1
        if award.expires_in <= 0:
            award.quality -= 1

    # fix any values that have exceeded quality parameters
    def fix_overages(self, award):
        if award.quality > 50: award.quality = 50
        if award.quality < 0: award.quality = 0

File: award.rb
class Award
  attr_accessor :name, :expires_in, :quality

  def initialize(name, expires_in, quality)
    @name       = name
    @expires_in = expires_in
    @quality    = quality
  end

  def update_quality(awards)
    awards.each do |award|
      # Blue Distinction Plus awards should not change
      if award.name != 'Blue Distinction Plus'
        case award.name
        when 'Blue First'
          update_bf(award)
        when 'Blue Compare'
          update_bc(award)
        when 'Blue Star'
          update_bs(award)
        else
          update_gen(award)
        end
        award.expires_in -= 1
        fix_overages(award)
      end
    end
  end

  # update Blue Compare award
  def update_bc(award)
    if award.expires_in > 0
      award.quality += 1
    end
    if award.expires_in <= 10
      award.quality += 1
    end
    if award.expires_in <= 5
      award.quality += 1
    end
    if award.expires_in <= 0
      award.quality = 0
    end
  end

  # update Blue First award
  def update_bf(award)
    award.quality += 1
    if award.expires_in <= 0
      award.quality += 1
    end
  end

  # update Blue Star award
  def update_bs(award)
    if award.expires_in > 0
      award.quality -= 2
    else
      award.quality -= 4
    end
  end

  # update all other awards
  def update_gen(award)
    award.quality -= 1
    if award.expires_in <= 0
      award.quality -= 1
    end
  end

  # fix any values that have exceeded quality parameters
  def fix_overages(award)
    if award.quality > 50
      award.quality = 50
    end
    if award.quality < 0
      award.quality = 0
    end
  end

end
